### ISSUE TITLE
[question] What form of package versions will be required for lockfile ci to work?

### ISSUE THREAD

**Issue by michaelmaguire** [conan-io/conan#8421]
In following through the https://docs.conan.io/en/latest/versioning/lockfiles/ci.html, we see that in the accompanying example packages https://github.com/conan-io/examples/tree/master/features/lockfiles/ci the package versions `requires=` is always specified using the "square bracket" https://docs.conan.io/en/latest/versioning/version_ranges.html e.g.:

```
class LibDConan(ConanFile):
    settings = "build_type"
    requires = "libb/[>0.0 <1.0]@user/testing", "libc/[>0.0  <1.0]@user/testing"
```

https://github.com/conan-io/examples/blob/master/features/lockfiles/ci/libd/conanfile.py#L8

So far in my team, we have been using the "overrideable" style of version specification:

e.g. 

```
    alias = "pkg/0.1@user/testing"
```

https://docs.conan.io/en/latest/versioning/introduction.html

Can I confirm that as we transition to using lockfile ci in our build, we'll need to change all our `requires=` package version specification from `pkg/0.1` to `pkg/[>=0.1]`?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @michaelmaguire 

No, it is not necessary to move to version ranges to work. It is also possible to use the plan requires, and use revisions (revisions is kind of an internal, automatic version-range that always resolve to latest revision, unless explicitly defined a fixed revision).

What cannot be tracked with lockfiles is if **not** using revision and not doing version bumps (which would be used by version ranges), because not using revision and not bumping the version basically overwrites the existing packages and artifacts. So there is no way to identify unique, immutable entities, and therefore, they cannot be locked.

I think that a potential popular flow would be using lockfiles with revisions and ``recipe_revision_mode``, to do aggressive, fast CI over massive concurrent developer changes that don't bump versions, and then manual version bumps, both in the packages and their consumers happen when things evolve in their interfaces and might expose (breaking or not) changes downstream.

Please let me know if this clarifies this issue. Thanks!

**Comment by ytimenkov** [comment]
@memsharded 

So this is the final recommendation?

> and then manual version bumps, both in the packages and their consumers

Particular problem with having fixed versions in recipes is that a new release is required for every new dependency propagation because recipe needs to be committed back to Git repo so previous version tag can no longer be used.

And if versions are stored in a lockfile it can't be updated in a controlled manner: say, change only one version to a specific one. `conan lock create` always goes to Artifactory to fetch a new version / revision or picks random one from cache.

We can't go with recipe revisions anyways because we're not ready yet and it's just unnatural in communication: "I have a bug in revision `4e90c236d9c0cb4ae785da475f2cd8b6c54d5d69` which was fixed in `134c0166b33125e661896380c5bf3fee3c353d66`. Is it still present in revision `d5b58644838e93ddb48f05b4a52ba862eff3c1b3`?" 🤯).

I was thinking that having version ranges and a lockfile on the side could be a good solution, similar to python's setup.py/requirements vs lockfiles: recipe defines compatibility requirements while lockfile particular configuration for the application.

Are there any principal obstacles why can `RangeResolver` or `GraphLock.lock_node` not take a version/revision from a supplied lockfile(s) instead of cache / remotes?

(I found it's related to #8331)
